File: leetcode/双指针/0019_Remove_Nth_Node_From_End_of_List/Solution.py
class Solution(object):
    def removeNthFromEnd(self, head, n):
        """
        :type head: ListNode
        :type n: int
        :rtype: ListNode
        """
        p = dummy = ListNode(-1, head)

        for _ in range(n):
            head = head.next

        while head:
            head = head.next
            p = p.next
        
        p.next = p.next.next

        return dummy.next


class ListNode(object):
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

File: leetcode/双指针/0019_Remove_Nth_Node_From_End_of_List/Solution.java
public class Solution {

    public ListNode removeNthFromEnd(ListNode head, int n) {
        ListNode dummy = new ListNode(-1, head);
        ListNode pre = dummy;

        // head 移动 n 位, 而不是 n + 1, 因为 pre 从 dummy开始
        for (int i = 0; i < n; i++) { // 不做 head 是否为 null 的判断, 因为题目保证 1 <= n <= sz
            head = head.next;
        }

        // 至此, head 已经就位, 开始移动 pre
        while (head != null) {
            head = head.next;
            pre = pre.next;
        }

        // pre 已经就位, 开始删除
        pre.next = pre.next.next;

        return dummy.next;
    }

    static class ListNode {
        int val;
        ListNode next;
        ListNode() {}
        ListNode(int val) { this.val = val; }
        ListNode(int val, ListNode next) { this.val = val; this.next = next; }
    }
}
